core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java

```diff
@@ -41,88 +41,128 @@
+    preserveOwner = conf.getBoolean(FileCopyOutputFormat.PRESERVE_OWNER, false);
+    fsCache = conf.getBoolean(FileCopyOutputFormat.FS_CACHE, true);
+    bufferSize = conf.getInt(FileCopyOutputFormat.BUFFER_SIZE, defaultBufferSize);
+    sourceFilesystemMap = new HashMap<>();
+  }
 
+  @Override
+  public void write(NullWritable key, AbstractFileMetadata fileMetadata) throws IOException, InterruptedException {
     // construct file paths for source and destination
-    Path srcPath = new Path(fileMetadata.fullPath);
-    Path destPath = new Path(basePath + "/" + fileMetadata.basePath);
+    Path srcPath = new Path(fileMetadata.getFullPath());
+    Path destPath;
+    if (fileMetadata.getBasePath().isEmpty()) {
+      // nothing to create
+      return;
+    } else {
+      destPath = new Path(basePath, fileMetadata.getBasePath());
+    }
     FsPermission permission = new FsPermission(fileMetadata.getPermission());
 
-    if (fileMetadata.isFolder) {
+    // immediately return if we don't want to overwrite and file exists in destination
+    if (!enableOverwrite && destFileSystem.exists(destPath)) {
+      return;
+    }
+
+    // get source database connection
+    String uriString = fileMetadata.getHostURI();
+    if (!sourceFilesystemMap.containsKey(uriString)) {
+      sourceFilesystemMap.put(uriString, getSourceFilesystemConnection(fileMetadata, fsCache));
+    }
+    FileSystem sourceFilesystem = sourceFilesystemMap.get(uriString);
+
+    if (fileMetadata.isFolder()) {
       // create an empty folder and return
       if (!destFileSystem.exists(destPath) && sourceFilesystem.isDirectory(srcPath)) {
         destFileSystem.mkdirs(destPath, permission);
+        if (preserveOwner) {
+          destFileSystem.setOwner(destPath, fileMetadata.getOwner(), fileMetadata.getGroup());
+        }
       }
       return;
     } else if (!sourceFilesystem.exists(srcPath)) {
       // file doesn't exist in source, return immediately
-      return;
-    }
-
-    // immediately return if we dont want to overwrite and file exists in destination
-    if (!enableOverwrite && destFileSystem.exists(destPath)) {
+      LOG.warn("{} doesn't exist in source filesystem.", fileMetadata.getFullPath());
       return;
     }
 
     // data streaming
     FSDataInputStream inputStream = sourceFilesystem.open(srcPath);
     FSDataOutputStream outputStream = FileSystem.create(destFileSystem, destPath, permission);
-    byte[] buf = new byte[bufferSize];
-    int len;
-    while ((len = inputStream.read(buf)) >= 0) {
-      outputStream.write(buf, 0, len);
+    try {
+      byte[] buf = new byte[bufferSize];
+      int len;
+      while ((len = inputStream.read(buf)) >= 0) {
+        outputStream.write(buf, 0, len);
+      }
+    } finally {
+      if (preserveOwner) {
+        destFileSystem.setOwner(destPath, fileMetadata.getOwner(), fileMetadata.getGroup());
+      }
+      inputStream.close();
```

Comment:
if this throws an exception, outputStream won't get closed. So you need to wrap this in a try catch finally as well.